src/match_identifiers.cpp

```diff
@@ -4,25 +4,37 @@
 #include <unordered_map>
 
 #include "compile_time_environment.hpp"
+#include "error_report.hpp"
 #include "typed_ast.hpp"
 
 #include <cassert>
 
 namespace TypeChecker {
 
-void match_identifiers(
+#define CHECK_AND_RETURN(expr)                                                 \
+	{                                                                          \
+		auto err = expr;                                                       \
+		if (!err.ok())                                                         \
+			return err;                                                        \
+	}
+
+[[nodiscard]] ErrorReport match_identifiers(
```

Comment:
`[[nodiscard]]` is a C++17 attribute. Although very useful, using it would mean that we should be discarding our home made `string_view.xpp`, because its main purpose was to lower us to C++14. We should decide on that.

Reply:
Not really. attributes are ignored by compilers that don't understand them. And some may provide the feature, even when compiling with older standards.

Reply:
I did not know that, huh